packages/fork-choice/src/protoArray/protoArray.ts

```diff
@@ -214,6 +232,135 @@ export class ProtoArray {
     }
   }
 
+  /**
+   * Optimistic sync validate till validated latest hash, invalidate any decendant branch if invalidate till hash provided
+   * TODO: implementation:
+   * 1. verify is_merge_block if the mergeblock has not yet been validated
+   * 2. Throw critical error and exit if a block in finalized chain gets invalidated
+   */
+  validateLatestHash(latestValidExecHash: RootHex, invalidateTillBlockHash: RootHex | null): void {
+    // We can skip calling this if the block was successfully validated and is
+    // to be imported, as on_block would already propagate valid hash
+
+    // Look reverse because its highly likely node with latestValidExecHash is towards the
+    // the leaves of the forkchoice
+    const latestValidHashIndexReverse = this.nodes
+      .slice()
```

Comment:
do we really need this `slice()` call?

Reply:
umm, may be i can  loop reverse (without slicing it off), would be efficient as well